component/examples/stream_example.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+extern crate component;
+
+use anyhow::anyhow;
+use async_trait::async_trait;
+use component::{IrrecoverableError, Manageable, Supervisor};
+use std::cmp::min;
+use std::sync::Once;
+use tokio::sync::mpsc::Sender;
+use tokio::sync::oneshot::Receiver as oneshotReceiver;
+use tokio::task::JoinHandle;
+
+static mut SHOULD_FAIL: bool = true;
+static FIXER: Once = Once::new();
+
+fn fix() {
+    FIXER.call_once(|| unsafe {
+        SHOULD_FAIL = false;
+    })
+}
+
+/// We create two structs, an empty struct as a component that will contain functions
+/// that create and fully encapsulate an instance of the actual type.
+/// An instance of the actual type needs to be instantiated inside the supervised task
+/// in order to have the correct lifetime.
+pub struct MockTcpStreamComponent {}
+
+pub struct MockTcpStream {
+    read_data: Vec<u8>,
+}
+
+impl MockTcpStream {
+    pub fn new() -> Self {
+        let read_data = Vec::new();
+        MockTcpStream { read_data }
+    }
+
+    /// This function will fail on the first call and then succeed on preceding calls to create
+    /// a situation where we have an irrecoverable error.
+    fn mock_read(&self, buf: &mut [u8]) -> Result<usize, anyhow::Error> {
+        // failure should happen once
+        unsafe {
+            if SHOULD_FAIL {
+                fix();
+                return Result::Err(anyhow!("Could not read from stream."));
+            }
+        }
+
+        let size: usize = min(self.read_data.len(), buf.len());
+        buf[..size].copy_from_slice(&self.read_data[..size]);
+        Ok(size)
+    }
+}
+
+impl MockTcpStreamComponent {
+    /// This is a function that should run continuously doing some operation, here we are
+    /// continuously listening on a mocked TCP Stream. This is ultimately the function that we
+    /// are supervising.
+    /// Inside this function we first initialize any state that will be used in this component so
+    /// that the scope is also correctly reset on a restart.
+    ///
+    /// This would be an excellent place to also add a scopeguard with a defer_panic so that if the
```

Comment:
+1 on the defer panic 👍

Reply:
Use of a scope guard is now completely obsolete since we will now pass a &mut self to handle_irrecoverable!